### PR TITLE
plugin/kubernetes: expose client-go internal request metrics

### PR DESCRIPTION
**Pull request by iyashu**
### 1. Why is this pull request needed and what does it do?
This pull request exposes following K8s client level metrics to monitor K8s requests latency & status codes. It is useful in setting up alerts (if required) in case coredns fails to query kube-apiserver (due to networking issues or application level issues) making it to serve stale records.
```
# HELP coredns_kubernetes_rest_client_rate_limiter_duration_seconds Client side rate limiter latency in seconds. Broken down by verb and URL.
# TYPE coredns_kubernetes_rest_client_rate_limiter_duration_seconds histogram
coredns_kubernetes_rest_client_rate_limiter_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="0.001"} 14
coredns_kubernetes_rest_client_rate_limiter_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="0.002"} 14
coredns_kubernetes_rest_client_rate_limiter_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="0.004"} 14
coredns_kubernetes_rest_client_rate_limiter_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="0.008"} 14
coredns_kubernetes_rest_client_rate_limiter_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="0.016"} 14
coredns_kubernetes_rest_client_rate_limiter_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="0.032"} 14
coredns_kubernetes_rest_client_rate_limiter_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="0.064"} 14
coredns_kubernetes_rest_client_rate_limiter_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="0.128"} 14
coredns_kubernetes_rest_client_rate_limiter_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="0.256"} 14
coredns_kubernetes_rest_client_rate_limiter_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="0.512"} 14
coredns_kubernetes_rest_client_rate_limiter_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="+Inf"} 14
coredns_kubernetes_rest_client_rate_limiter_duration_seconds_sum{url="10.24.1.228:6443",verb="GET"} 0.000188625
coredns_kubernetes_rest_client_rate_limiter_duration_seconds_count{url="10.24.1.228:6443",verb="GET"} 14

# HELP coredns_kubernetes_rest_client_request_duration_seconds Request latency in seconds. Broken down by verb and URL.
# TYPE coredns_kubernetes_rest_client_request_duration_seconds histogram
coredns_kubernetes_rest_client_request_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="0.001"} 0
coredns_kubernetes_rest_client_request_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="0.002"} 0
coredns_kubernetes_rest_client_request_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="0.004"} 0
coredns_kubernetes_rest_client_request_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="0.008"} 0
coredns_kubernetes_rest_client_request_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="0.016"} 0
coredns_kubernetes_rest_client_request_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="0.032"} 0
coredns_kubernetes_rest_client_request_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="0.064"} 0
coredns_kubernetes_rest_client_request_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="0.128"} 4
coredns_kubernetes_rest_client_request_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="0.256"} 6
coredns_kubernetes_rest_client_request_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="0.512"} 8
coredns_kubernetes_rest_client_request_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="1.024"} 8
coredns_kubernetes_rest_client_request_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="2.048"} 8
coredns_kubernetes_rest_client_request_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="4.096"} 8
coredns_kubernetes_rest_client_request_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="8.192"} 8
coredns_kubernetes_rest_client_request_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="16.384"} 8
coredns_kubernetes_rest_client_request_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="32.768"} 14
coredns_kubernetes_rest_client_request_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="65.536"} 14
coredns_kubernetes_rest_client_request_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="131.072"} 14
coredns_kubernetes_rest_client_request_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="262.144"} 14
coredns_kubernetes_rest_client_request_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="524.288"} 14
coredns_kubernetes_rest_client_request_duration_seconds_bucket{url="10.24.1.228:6443",verb="GET",le="+Inf"} 14
coredns_kubernetes_rest_client_request_duration_seconds_sum{url="10.24.1.228:6443",verb="GET"} 181.56279254400002
coredns_kubernetes_rest_client_request_duration_seconds_count{url="10.24.1.228:6443",verb="GET"} 14

# HELP coredns_kubernetes_rest_client_requests_total Number of HTTP requests, partitioned by status code, method, and host.
# TYPE coredns_kubernetes_rest_client_requests_total counter
coredns_kubernetes_rest_client_requests_total{code="200",host="10.24.1.228:6443",method="GET"} 14
coredns_kubernetes_rest_client_requests_total{code="<error>",host="10.24.1.228:6443",method="GET"} 6
```
You may refer to [this metrics initialisation](https://github.com/kubernetes/kubernetes/blob/0f373abb6a0de9a8a7eee14462701eb2875ba9d6/staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go#L34) pkg to see how K8s internal components (scheduler, controller-manager etc.) expose these client level metrics.
### 2. Which issues (if any) are related?
NA
### 4. Does this introduce a backward incompatible change or deprecation?
No.

### CODE REVIEW

File: plugin/kubernetes/metrics.go
Comment (Tantalor93) [inline plugin/kubernetes/metrics.go]: why are you calling it a Summary if it is a Histogram? https://prometheus.io/docs/practices/histograms/

File: plugin/kubernetes/metrics.go
Comment (Tantalor93) [inline plugin/kubernetes/metrics.go]:
```suggestion
func (l *latencyAdapter) Observe(_ context.Context, verb string, u url.URL, latency time.Duration) {
```

File: plugin/kubernetes/metrics.go
Comment (Tantalor93) [inline plugin/kubernetes/metrics.go]:
```suggestion
func (r *resultAdapter) Increment(_ context.Context, code, method, host string) {
```

File: plugin/kubernetes/metrics.go
Comment (Tantalor93) [inline plugin/kubernetes/metrics.go]: I think that using `prometheus.DefBuckets` would be better, or why do you think that these custom buckets are better?

File: plugin/kubernetes/metrics.go
Comment (Tantalor93) [inline plugin/kubernetes/metrics.go]: use default buckets?

File: plugin/kubernetes/metrics.go
Comment (Tantalor93) [reply plugin/kubernetes/metrics.go]:
```suggestion
	// requestLatency measures latency of k8s client requests grouped by verb and url.
```

File: plugin/kubernetes/metrics.go
Comment (Tantalor93) [inline plugin/kubernetes/metrics.go]: not a summary
```suggestion
	// rateLimiterLatency measures rest rate limiter latency grouped by verb and url.
```

**Review by Tantalor93 — CHANGES_REQUESTED** [review]

File: plugin/kubernetes/metrics.go
Comment (iyashu) [reply plugin/kubernetes/metrics.go]: I was not aware of default buckets from prometheus library. May be we can start with the same & see in future if we require to capture additional buckets.

**Comment by iyashu** [comment]
@Tantalor93 thanks for reviewing this pull request. I've addressed all of the comments. ptal.

**Comment by iyashu** [comment]
@Tantalor93  @chrisohaver gentle reminder :)

**Review by Tantalor93 — DISMISSED** [review]
LGTM, thanks! Let's give it few more days before merge for other maintainers to have chance to review

**Review by Tantalor93 — COMMENTED** [review]
hmm, noticed that we forgot to update README with new metrics, can you please add them? https://github.com/coredns/coredns/blob/master/plugin/kubernetes/README.md#metrics

**Comment by SuperQ** [comment]
Nice. Can you explain more about what the rate limit latency is?

**Comment by iyashu** [comment]
> hmm, noticed that we forgot to update README with new metrics, can you please add them? https://github.com/coredns/coredns/blob/master/plugin/kubernetes/README.md#metrics

@Tantalor93 thanks for the revert. I've updated the README capturing documentation for these metrics. ptal.

**Comment by iyashu** [comment]
>Can you explain more about what the rate limit latency is?

@SuperQ it is the [latency contributed](https://github.com/kubernetes/client-go/blob/master/rest/request.go#L635) by client side rate limiter [enabled by default](https://github.com/kubernetes/client-go/blob/master/rest/config.go#L353-L363) in client-go. It is useful to measure delays in sending request to apiserver due to throttling by client side rate limiter.

**Review by Tantalor93 — APPROVED** [review]
lgtm

**Comment by iyashu** [comment]
@chrisohaver @SuperQ gentle reminder.

File: plugin/kubernetes/README.md
Comment (chrisohaver) [inline plugin/kubernetes/README.md]:
```suggestion
The following are client level metrics to monitor apiserver request latency & status codes. `verb` identifies the apiserver [request type](https://kubernetes.io/docs/reference/using-api/api-concepts/#single-resource-api) and `url`/`host` denotes the apiserver endpoint.
```

File: plugin/kubernetes/README.md
Comment (chrisohaver) [inline plugin/kubernetes/README.md:232]:
```suggestion

The following are client level metrics to monitor apiserver request latency & status codes. `verb` identifies the apiserver [request type](https://kubernetes.io/docs/reference/using-api/api-concepts/#single-resource-api) and `url`/`host` denotes the apiserver endpoint.
```

**Review by chrisohaver — CHANGES_REQUESTED** [review]

**Review by chrisohaver — APPROVED** [review]

**Review by Tantalor93 — APPROVED** [review]

**Comment by SuperQ** [comment]
I'm not sure `url` is the correct label name. Isn't this the `apiserver`? Also `verb` looks more like `method`, as it seems to be an HTTP request method.

**Comment by iyashu** [comment]
> I'm not sure url is the correct label name. Isn't this the apiserver? Also verb looks more like method, as it seems to be an HTTP request method.

@SuperQ `verb` is commonly used term in K8s representing more information on request type. See [this](https://kubernetes.io/docs/reference/using-api/api-concepts/#single-resource-api) for more details. But I think it makes sense to rename `url` as `host` as we're extracting `host` under [latency adapter](https://github.com/coredns/coredns/pull/5991/files#diff-4cf2403647f813a5142b52c727596ae3c5983e359e8f38057001d3329f11c98aR65) rather than actual request url to avoid cardinality explosion. Raised https://github.com/coredns/coredns/pull/6040 to rename the label name. Thanks.

**Pull request merged**
